sdk/mobile-center-updates/src/main/java/com/microsoft/azure/mobile/updates/Updates.java

```diff
@@ -0,0 +1,971 @@
+    }
+
+    /**
+     * Implements {@link #setApiUrl(String)}}.
+     */
+    private synchronized void setInstanceApiUrl(String apiUrl) {
+        mApiUrl = apiUrl;
+    }
+
+    /**
+     * Cancel everything.
+     */
+    private synchronized void cancelPreviousTasks() {
+        if (mCheckReleaseApiCall != null) {
+            mCheckReleaseApiCall.cancel();
+            mCheckReleaseApiCall = null;
+            mCheckReleaseCallId = null;
+        }
+        mUpdateDialog = null;
+        mReleaseDetails = null;
+        if (mDownloadTask != null) {
+            mDownloadTask.cancel(true);
+            mDownloadTask = null;
+        }
+        if (mProcessDownloadCompletionTask != null) {
+            mProcessDownloadCompletionTask.cancel(true);
+            mProcessDownloadCompletionTask = null;
+        }
+        long downloadId = StorageHelper.PreferencesStorage.getLong(PREFERENCE_KEY_DOWNLOAD_ID);
+        if (downloadId > 0) {
+            MobileCenterLog.debug(LOG_TAG, "Removing download and notification id=" + downloadId);
+            removeDownload(downloadId);
+        }
+        StorageHelper.PreferencesStorage.remove(PREFERENCE_KEY_DOWNLOAD_ID);
+        StorageHelper.PreferencesStorage.remove(PREFERENCE_KEY_DOWNLOAD_URI);
+    }
+
+    /**
+     * Method that triggers the update workflow or proceed to the next step.
+     */
+    private synchronized void resumeUpdateWorkflow() {
+        if (mForegroundActivity != null && !mWorkflowCompleted && isInstanceEnabled()) {
+
+            /* If we received the update token before Mobile Center was started/enabled, process it now. */
+            if (mBeforeStartUpdateToken != null) {
+                MobileCenterLog.debug(LOG_TAG, "Processing update token we kept in memory before onStarted");
+                storeUpdateToken(mBeforeStartUpdateToken, mBeforeStartRequestId);
+                mBeforeStartUpdateToken = null;
+                mBeforeStartRequestId = null;
+                return;
+            }
+
+            /* If we have a download ready but we were in background, pop install UI now. */
+            String downloadUri = StorageHelper.PreferencesStorage.getString(PREFERENCE_KEY_DOWNLOAD_URI);
+            if ("".equals(downloadUri)) {
+
+                /* TODO double check that with download manager. */
+                MobileCenterLog.verbose(LOG_TAG, "Download is still in progress...");
+                return;
+            } else if (downloadUri != null)
+                try {
+
+                    /* FIXME this can cause strict mode violation. */
+                    Uri apkUri = Uri.parse(downloadUri);
+                    MobileCenterLog.debug(LOG_TAG, "Now in foreground, remove notification and start install for APK uri=" + apkUri);
+                    mForegroundActivity.startActivity(getInstallIntent(apkUri));
+                    NotificationManager notificationManager = (NotificationManager) mContext.getSystemService(Context.NOTIFICATION_SERVICE);
```

Comment:
We have multiple entry points in this class as it can be called in background or via onStart, and we need context.